### ISSUE TITLE
Allow the metric expression to include columns that come from linked Dimension node

### ISSUE THREAD

**Issue by anhqle** [DataJunction/dj#1188]
Currently, metric expression can only include columns from the Transform node.

Example of why this would be useful:
1. user_latest_status = `MAX_BY(user_status, time_column_from_dim_node)`
2. engagement_on_android = `SUM(engagement) FILTER (WHERE device_name_from_dim_node = 'android')`

**Comment by samredai** [comment]
Hmm, that would mean the column from the dim_node is required and so it's reasonable to require setting dim_node as a required dimension beforehand right? That would help limit the validation so we don't have to search the entire dimension dag to find this column. We could just look at the attributes (columns) on the dimensions found in the required dimensions for that metric.

Which brings me to another question. What if multiple dimension nodes have that particular column name? Do we just fail validation or SQL generation and raise some error indicating an ambiguous value in the metric definition?

**Comment by anhqle** [comment]
> that would mean the column from the dim_node is required and so it's reasonable to require setting dim_node as a required dimension beforehand right?

If by `required dimension`, you mean colloquially that the dim node must be joined for the metric to be computable, then yes, it is required.

If by `required dimension` you mean the DJ-definition i.e. "a grain at which the metric has to be consumed" (this name is frankly still unintuitive to me...), then no.

> What if multiple dimension nodes have that particular column name? Do we just fail validation or SQL generation and raise some error indicating an ambiguous value in the metric definition?

Failing seems fine to me. 
Requiring the user to write the full column name could also be okay, except that it'd be weird for the Transform's column to not be fully qualified, but the Dimension's column is.

**Comment by samredai** [comment]
Sorry for the late reply! GitHub notifications are....well.

> If by `required dimension`, you mean colloquially that the dim node must be joined for the metric to be computable, then yes, it is required.

I meant that the metric has the dimension attribute in question listed as a "required_dimension" on the metric definition. As in [this](https://github.com/DataJunction/dj/blob/main/datajunction-clients/python/datajunction/nodes.py#L505) value being populated via whichever client that created/updated the node (that example is using the python client).

I'm unsure if that's sufficient for your second example `engagement_on_android`: `SUM(engagement) FILTER (WHERE device_name_from_dim_node = 'android')`. That almost looks like a separate concept of a required filter which currently doesn't exist but is something we could potentially add. Is your expectation that the dimension node would be filtered on `device_name_from_dim_node` first and _then_ joined to metric (meaning the underlying transform node)? If so then we do carry around a context during the SQL generation so if we allowed that context to carry a bunch of these "required filters" from metric definitions, then it can apply it only when that dimension is pulled in. If you expect that dimension to **always** be pulled in then you could separately set it as a required dimension.

**Comment by anhqle** [comment]
> Is your expectation that the dimension node would be filtered on device_name_from_dim_node first and then joined to metric (meaning the underlying transform node)?

I don't mind this, but it's also not necessary to filter the dimension node first.

IIUC the difficulty, basically DJ needs to somehow know that it has to perform the join with dim node so that `device_name_from_dim_node` can be used in the metric expression? That makes sense, but for a v0, I'm also okay with putting that onus on the client. In other words, the client needs to request all the necessary dimension to be used in the metric expression. If the client fails to do so, then the generated query can fail at run time.

**Comment by samredai** [comment]
> And that your point about `required_dimension` is basically a way for DJ to know that?

Yeah that's what I'm thinking. If it's a required dimension, then we've validated that the dimension attribute is available (via a join to the dimension) and also that the join will always happen, hence "required". That's basically all we need to know that the metric expression is valid.

> I'm also okay with putting that onus on the client. In other words, the client needs to request all the necessary dimension to be used in the metric expression. If the client fails to do so, then the generated query fails at run time.

My hesitation with leaving it to the client (as opposed to having the metric author add a required dimension) is that we would basically have no way for the backend system to truly know if the metric definition is valid until runtime. When the user creates/updates the metric, we'll validate that the column in the metric expression exists in the underlying transform node and then give a nice green checkmark and just assume any columns not found must be somewhere out in the dimension graph--so when the client gets an error for 100% of requests for that metric, it could be confusing.

**Comment by shangyian** [comment]
> I meant that the metric has the dimension attribute in question listed as a "required_dimension" on the metric definition

I actually don't think that putting it as a required dimension makes sense in this case, because required dimensions imply that the metric can only be interpreted when it's grouped by those dimensions (i.e., the metric only makes sense at the level of those required dimensions). I don't think we should reuse this concept to represent a different thing, because that adds confusion. Although arguably the term "required dimension" is already confusing...

> IIUC the difficulty, basically DJ needs to somehow know that it has to perform the join with dim node so that device_name_from_dim_node can be used in the metric expression?

DJ already has the ability to detect what dimensions are available for a given node, so it's more that we need to do the following:
* Relax the upstream references of a metric to include both the upstream node *and* its dimensions
* Whenever SQL for this metric is generated, automatically do the join to pull in the referenced dimension
